### PR TITLE
[#9,#14,#17,irods/irods#7265] Minor reorganization + CMake TLC

### PR DESCRIPTION
**Pull request by SwooshyCueb**
In service of #9
In service of irods/irods#7265
Addresses #14
Addresses #17
Supersedes #20

- removed unused `main.cpp`
- reorganized project
- created object library target for `auth_check_wrapper.cpp`
- created interface library target for `ipam_client.hpp`
- quoted paths in cmake
- Removed unneeded `CMAKE_SHARED_LINKER_FLAGS_INIT` and `CMAKE_SHARED_LINKER_FLAGS_RELEASE_INIT`
- Added missing `CMAKE_EXE_LINKER_FLAGS_INIT` and `CMAKE_EXE_LINKER_FLAGS_RELEASE_INIT`
- Swapped quotes to angle brackets on irods includes
- Removed references to nonexistent include directory from cmake
- Removed unneeded `pam_conversation.hpp` include from `handshake_session.cpp`
- Bumped irods version and removed exact requirement
- Updated `CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION`
- Split into server and client packages (server package depends on client package)
- Replaced PAM `find_library` logic with a new find module
- Switched to `find_package` for fmt and spdlog

### CODE REVIEW

File: cmake/Modules/FindPAM.cmake [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+``PAM_FOUND``
+  true if PAM headers and library were found
+``PAM_LIBRARY``
+  PAM library to be linked
+``PAM_INCLUDE_DIR``
+  the directory containing PAM headers
+
+TODO
+^^^^
```
Comment (korydraughn) [inline cmake/Modules/FindPAM.cmake:28]: Issue number?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,194 +1,95 @@
...
-#set(CPACK_RPM_POST_INSTALL_SCRIPT_FILE "${CMAKE_SOURCE_DIR}/packaging/postinst")
 set(CPACK_RPM_FILE_NAME RPM-DEFAULT)
 
-set(CPACK_ARCHIVE_COMPONENT_INSTALL OFF)
+string(TOUPPER "${project_component_part}-client" irods_plugin_package_client_component)
+string(TOUPPER "${project_component_part}-server" irods_plugin_package_server_component)
+
+set(CPACK_DEBIAN_${irods_plugin_package_client_component}_PACKAGE_NAME "${PLUGIN_PACKAGE_NAME}-client")
+set(CPACK_DEBIAN_${irods_plugin_package_client_component}_PACKAGE_DEPENDS "irods-runtime (= ${IRODS_VERSION}), libc6")
+set(CPACK_DEBIAN_${irods_plugin_package_client_component}_PACKAGE_BREAKS "${PLUGIN_PACKAGE_NAME} (<< 4.90.0.0-1~), irods-auth-interactive-pam (<< 4.90.0.0-1~)")
```
Comment (korydraughn) [inline CMakeLists.txt:76]: Is `irods-auth-interactive-pam` correct?
Is it supposed to be `irods-auth-plugin-pam-interactive`?

Comment (SwooshyCueb) [reply CMakeLists.txt:76]: Yes, `irods-auth-interactive-pam` was the name prior to https://github.com/irods/irods_auth_plugin_pam_interactive/commit/8bfa0f6be084be39cf7f0f849f2e6f897913f5f4.
`PLUGIN_PACKAGE_NAME` is `irods-auth-plugin-pam-interactive`

File: cmake/Modules/FindPAM.cmake (hunk 1 above)
Comment (SwooshyCueb) [reply cmake/Modules/FindPAM.cmake:28]: irods/irods#6247

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,194 +1,95 @@
...
 include(GNUInstallDirs)
 include(UseLibCXX)
 
 find_package(Threads REQUIRED)
 find_package(OpenSSL REQUIRED COMPONENTS Crypto SSL)
 find_package(nlohmann_json "3.6.1" REQUIRED)
+	find_package(fmt "8.1.1"
+		HINTS "${IRODS_EXTERNALS_FULLPATH_FMT}")
+	find_package(spdlog "1.9.2"
+		HINTS "${IRODS_EXTERNALS_FULLPATH_SPDLOG}")
```
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply CMakeLists.txt]: Nope, fixed

File: pam_handshake_auth_check/CMakeLists.txt
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply pam_handshake_auth_check/CMakeLists.txt:1]: Fixed

**Comment by SwooshyCueb** [comment]
Now that we've got 4.3.2 out, I think this PR is ready for review

File: cmake/Modules/FindPAM.cmake [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+include(FindPackageHandleStandardArgs)
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(
+	PAM
+	REQUIRED_VARS PAM_INCLUDE_DIR PAM_LIBRARY
+)
+
+if (PAM_FOUND)
+	if (NOT TARGET PAM::pam)
```
Comment (korydraughn) [inline cmake/Modules/FindPAM.cmake:57]: Why is the casing for `PAM::pam` different?
Aren't they normally the same?
Or are we dodging a naming collision?

**Review by korydraughn — COMMENTED** [review]
Looks good overall. Spotted one thing though.

File: cmake/Modules/FindPAM.cmake (hunk 4 above)
Comment (SwooshyCueb) [reply cmake/Modules/FindPAM.cmake:57]: > Aren't they normally the same?

Ehhh, it's not a convention.

The name and case of the first part typically has to match the name of the find module. I decided to keep the target name itself lowercase for consistency, as there are other libraries that are part of the pam package that are almost always typset in lowercase, and we may want to implement support for them at some point.

**Review by korydraughn — APPROVED** [review]
Let's get one more approval before pounding.

**Review by alanking — APPROVED** [review]
Looks good to me. I say we get this in.

Even though this is unreleased, it may be in our interest to notify our known users (particularly the Consortium members) about this change at it will make the plugin incompatible with older iRODS server versions moving forward. Not sure how we do that, but just gonna poke @korydraughn.

**Comment by korydraughn** [comment]
Incompatible due to the CMake changes?

**Comment by alanking** [comment]
Eh, perhaps that just comes with the territory of building something yourself. Just wanted to make sure we considered it.

**Comment by korydraughn** [comment]
I see. Given this project is external, we adjust it so that it compiles against 4.3.1 and later.

I think it would be trivial given we did that with the GenQuery2 repo.

@SwooshyCueb What would it take to make this compile against 4.3.1 too? Is it just adjusting the minimum iRODS version?

**Comment by SwooshyCueb** [comment]
> @SwooshyCueb What would it take to make this compile against 4.3.1 too? Is it just adjusting the minimum iRODS version?

I think it may require some conditional logic around fmt and spdlog handling in CMake. I'm not sure though.

**Comment by korydraughn** [comment]
Okay. Let's make that happen so people aren't required to upgrade to 4.3.2.

**Comment by SwooshyCueb** [comment]
> Okay. Let's make that happen so people aren't required to upgrade to 4.3.2.

Roger that. It might take me a bit though as I don't currently have a workflow for building against 4.3.1.

**Comment by SwooshyCueb** [comment]
I've restored build compatibility with 4.3.1.
If there are no objections, I will \#.

**Comment by korydraughn** [comment]
Yep, squash and pound to taste.

**Comment by korydraughn** [comment]
Please make sure the issue numbers are correct too.

**Comment by SwooshyCueb** [comment]
\#'d

**Pull request merged**